File: pupa_franchise/pupa_franchise/report/pupa_stock_balance_summary/pupa_stock_balance_summary.js
// Copyright (c) 2026, Tridots and contributors
// For license information, please see license.txt

frappe.query_reports["Pupa Stock Balance Summary"] = {
	filters: [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			reqd: 1,
			on_change: function () {
				let company = frappe.query_report.get_filter_value("company");
				let warehouse_filter = frappe.query_report.get_filter("warehouse");
				if (!warehouse_filter) return;

				if (company) {
					frappe.call({
						method: "pupa_franchise.api.api_sync.get_pupa_warehouses",
						args: { company: company },
						callback: function (r) {
							if (r && r.message && r.message.length) {
								let warehouse_options = [""].concat(r.message);
								warehouse_filter.df.options = warehouse_options.join("\n");
								warehouse_filter.refresh();
								frappe.query_report.set_filter_value("warehouse", "");
							} else {
								warehouse_filter.df.options = "\n";
								warehouse_filter.refresh();
								frappe.query_report.set_filter_value("warehouse", "");
								frappe.throw(__("No warehouse is mapped against this company in Pupa Settings"));
							}
						}
					});
				} else {
					warehouse_filter.df.options = "\n";
					warehouse_filter.refresh();
					frappe.query_report.set_filter_value("warehouse", "");
				}
			}
		},
		{
			fieldname: "warehouse",
			label: __("Warehouse"),
			fieldtype: "Select",
			options: "",
			reqd: 1,
			on_change: function () {
				frappe.query_report.refresh();
			}
		}
	],

	onload: function (report) {
		let company = frappe.query_report.get_filter_value("company");
		if (company) {
			frappe.query_report.get_filter("company").on_change();
		}
	},

	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (column.fieldname === "actual_qty" && data && data.actual_qty > 0) {
			value = "<span style='color:green; font-weight:bold'>" + value + "</span>";
		}
		return value;
	}
};

File: pupa_franchise/pupa_franchise/report/pupa_stock_balance_summary/pupa_stock_balance_summary.py
# Copyright (c) 2026, Tridots and contributors
# For license information, please see license.txt

import frappe
import json
import requests


def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data


def get_columns():
    return [
        {
            "fieldname": "item_code",
            "label": "Item Code",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "fieldname": "item_name",
            "label": "Item Name",
            "fieldtype": "Data",
            "width": 200
        },
        {
            "fieldname": "item_group",
            "label": "Item Group",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "fieldname": "warehouse",
            "label": "Warehouse",
            "fieldtype": "Data",
            "width": 180
        },
        {
            "fieldname": "stock_uom",
            "label": "Stock UOM",
            "fieldtype": "Data",
            "width": 100
        },
        {
            "fieldname": "actual_qty",
            "label": "Actual Qty",
            "fieldtype": "Float",
            "width": 120
        },
        {
            "fieldname": "valuation_rate",
            "label": "Valuation Rate",
            "fieldtype": "Currency",
            "width": 130
        },
        {
            "fieldname": "stock_value",
            "label": "Stock Value",
            "fieldtype": "Currency",
            "width": 150
        }
    ]


def get_data(filters):
    warehouse = filters.get("warehouse")
    if not warehouse:
        return []

    try:
        from pupa_franchise.api.api_sync import get_api_settings
        base_url, headers = get_api_settings()

        response = requests.get(
            url=f"{base_url}/api/method/pupa.api.franchise.get_warehouse_stock_balance",
            headers=headers,
            params={"warehouse": warehouse}
        )

        if response.status_code != 200:
            frappe.msgprint(
                f"Error fetching stock from HQ. Status: {response.status_code}",
                indicator="red",
                alert=True
            )
            return []

        result = response.json().get("message", [])

        # Apply client-side filters
        item_group = filters.get("item_group")
        item_code = filters.get("item_code")

        if item_group:
            result = [r for r in result if r.get("item_group") and item_group.lower() in r["item_group"].lower()]

        if item_code:
            result = [r for r in result if r.get("item_code") and item_code.lower() in r["item_code"].lower()]

        return result

    except Exception as e:
        frappe.log_error(
            message=frappe.get_traceback(),
            title="Pupa Stock Balance Summary Error"
        )
        frappe.msgprint(
            f"Error fetching stock data from HQ: {str(e)}",
            indicator="red",
            alert=True
        )
        return []
